File: Ch04/04_02/sim.py
from dataclasses import dataclass


@dataclass
class Location:
    lat: int
    lng: int

    def __post_init__(self):
        if self.lat < -90 or self.lat > 90:
            raise ValueError(f'invalid lat: {self.lat}')
        if self.lng < -180 or self.lng > 180:
            raise ValueError(f'invalid lng: {self.lng}')

    def move(self, lat, lng):
        self.lat = lat
        self.lng = lng


if __name__ == '__main__':
    loc = Location(lat=32.5253837, lng=34.9427434)
    loc.move(32.0641339, 34.8742343)
    print(loc)

File: Ch04/04_02/sim.go
package main

import "fmt"

type Location struct {
	Lat float64
	Lng float64
}

func NewLocation(lat, lng float64) (Location, error) {
	if lat < -90 || lat > 90 {
		return Location{}, fmt.Errorf("invalid lat: %#v", lat)
	}
	if lng < -180 || lng > 180 {
		return Location{}, fmt.Errorf("invalid lng: %#v", lng)
	}

	loc := Location{
		Lat: lat,
		Lng: lng,
	}
	return loc, nil
}

func (l Location) Move(lat, lng float64) {
	l.Lat = lat
	l.Lng = lng
}

func main() {
	loc, err := NewLocation(32.5253837, 34.9427434)
	if err != nil {
		fmt.Println("ERROR:", err)
		return
	}
	loc.Move(32.0641339, 34.8742343)
	fmt.Printf("%#v\n", loc)
}
